morebits.js

```diff
@@ -1308,29 +1308,239 @@ Morebits.unbinder.getCallback = function UnbinderGetCallback(self) {
+	},
+	getUTCMonthNameAbbrev: function() {
+		return Morebits.date.localeData.monthsShort[this.getUTCMonth() + 1];
+	},
+	getMonthName: function() {
+		return Morebits.date.localeData.months[this.getMonth() + 1];
+	},
+	getMonthNameAbbrev: function() {
+		return Morebits.date.localeData.monthsShort[this.getMonth() + 1];
+	},
+	getUTCDayName: function() {
+		return Morebits.date.localeData.days[this.getUTCDay()];
+	},
+	getUTCDayNameAbbrev: function() {
+		return Morebits.date.localeData.daysShort[this.getUTCDay()];
+	},
+	getDayName: function() {
+		return Morebits.date.localeData.days[this.getDay()];
+	},
+	getDayNameAbbrev: function() {
+		return Morebits.date.localeData.daysShort[this.getDay()];
+	},
+
+	/**
+	 * Add a given number of minutes, hours, days, months or years to the date.
+	 * This is done in-place. The modified date object is also returned, allowing chaining.
+	 * @param {number} number
+	 * @param {string} unit
+	 * @throws {Error} if invalid or unsupported unit is given
+	 * @returns {Morebits.date}
+	 */
+	add: function(number, unit) {
+		// mapping time units with getter/setter function names
+		var unitMap = {
+			minutes: 'Minutes',
+			hours: 'Hours',
+			days: 'Date',
+			months: 'Month',
+			years: 'FullYear'
+		};
+		var unitNorm = unitMap[unit] || unitMap[unit + 's']; // so that plural forms also work
+		if (unitNorm) {
+			this['set' + unitNorm](this['get' + unitNorm]() + number);
+			return this;
+		}
+		throw new Error('Invalid unit "' + unit + '": Only ' + Object.keys(unitMap).join(', ') + ' are allowed.');
+	},
+
+	/**
+	 * Subtracts a given number of minutes, hours, days, months or years to the date.
+	 * This is done in-place. The modified date object is also returned, allowing chaining.
+	 * @param {number} number
+	 * @param {string} unit
+	 * @throws {Error} if invalid or unsupported unit is given
+	 * @returns {Morebits.date}
+	 */
+	subtract: function(number, unit) {
+		return this.add(-number, unit);
+	},
+
+	/**
+	 * Formats the date into a string per the given format string.
+	 * Replacement syntax is a subset of that in moment.js.
+	 * @param {string} formatstr
+	 * @param {(string|number)} [zone=system] - 'system' (for browser-default time zone), 'utc' (for UTC),
+	 * 'user' (for time zone according to user's site preferences), or specify a time zone as number of
+	 * minutes past UTC.
+	 * If using the zone=user, note that (i) mediawiki.user dependency is reuqired, (ii) it doesn't account
+	 * for changing of timezone offsets due to DST.
+	 * @returns {string}
+	 */
+	format: function(formatstr, zone) {
+		var udate = this;
+		// create a new date object that will contain the date to display as system time
+		if (zone === 'utc') {
+			udate = new Morebits.date(this.getTime()).add(this.getTimezoneOffset(), 'minutes');
+		} else if (zone === 'user') {
+			var offset = parseInt(mw.user.options.get('timecorrection').split('|')[1]);
+			udate = new Morebits.date(this.getTime()).add(offset, 'minutes');
```

Comment:
Thanks for taking the time out to research this, though I only needed to read the first line of your comment to know what the trouble was.

Reply:
In fact, I think there isn't much utility to having this because of the DST bug. MediaWiki doesn't expose the DST-adjusted timezone offset in the JS interface. Its available only for the PHP side of things. JJMC89 shows the issue succinctly in [phab:T223002#5843155](https://phabricator.wikimedia.org/T223002#5843155). 

Extensions that need to show timestamps using JS appear to be using the system time only (eg: [MediaViewer](https://github.com/wikimedia/mediawiki-extensions-MultimediaViewer/blob/master/resources/mmv/ui/mmv.ui.metadataPanel.js#L809)).

Reply:
👍